optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/common/BavetNodeBuildPolicy.java

```diff
@@ -77,4 +86,22 @@ public int getNodeOrderMaximum() {
         return joinConstraintStreamToJoinBridgeNodeMap;
     }
 
+    public List<BavetNode> getCreatedNodes() {
+        // Make a sequential list of unique nodes.
+        SortedMap<Integer, BavetNode> nodeIndexToNodeMap = Stream.concat(sharableNodeMap.keySet().stream(),
+                constraintIdToScoringNodeMap.values().stream())
+                .distinct()
+                .collect(Collectors.toMap(k -> k.getNodeIndex(), Function.identity(), (a, b) -> {
+                    throw new IllegalStateException("Impossible state: nodes (" + a + ") and (" + b + ") share index.");
+                }, TreeMap::new));
+        // Ensure there are no gaps in that list.
+        int maxNodeIndex = nodeIndexToNodeMap.lastKey();
+        int expectedMaxNodeIndex = nodeIndexToNodeMap.size() - 1;
+        if (maxNodeIndex != expectedMaxNodeIndex) {
+            throw new IllegalStateException("Impossible state: maximum node index (" + maxNodeIndex +
+                    ") does not match the expected maximum node count (" + expectedMaxNodeIndex + ").");
+        }
+        return Collections.unmodifiableList(new ArrayList<>(nodeIndexToNodeMap.values()));
+    }
```

Comment:
This is heavy. Thankfully it is only ever used once per `BavetConstraintSession`.
Considering that it fails fast on a pretty important thing, I say we need it.

Reply:
100% agree that is worth it.